torch/csrc/distributed/rpc/rpc_with_autograd.cpp

```diff
@@ -0,0 +1,161 @@
+#include <torch/csrc/distributed/rpc/rpc_with_autograd.h>
+#include <torch/csrc/byte_order.h>
+#include <torch/csrc/distributed/rpc/utils.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+constexpr int kAutogradMessageSize = 17;
+
+AutogradMetadata::AutogradMetadata(
+    int64_t autogradContextId_,
+    int64_t autogradMessageId_)
+    : autogradContextId(autogradContextId_),
+      autogradMessageId(autogradMessageId_) {}
+
+RpcWithAutograd::RpcWithAutograd(
+    MessageType messageType,
+    const AutogradMetadata& autogradMetadata,
+    std::unique_ptr<RpcCommandBase> wrappedRpc)
+    : messageType_(messageType),
+      autogradMetadata_(autogradMetadata),
+      wrappedRpc_(std::move(wrappedRpc)) {
+  TORCH_INTERNAL_ASSERT(wrappedRpc_ != nullptr, "wrappedRpc cannot be null!");
+  TORCH_INTERNAL_ASSERT(
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_REQ ||
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_RESP);
+  wrappedMessage_ = wrappedRpc_->toMessage();
```

Comment:
As I read the patch, I find myself increasingly surprised by the fact that autograd messages wrap normal messages. Here's why.

Historically, in PyTorch proper, we indeed had a "wrapper" abstraction for Variable and Tensor. A Variable was a wrapper over Tensors which implements autograd. However, over time, we realized that wrapping this way was a mistake, because people continually made mistakes about whether or not they had a Variable or a Tensor, and the fact that a non-autograd recording Variable was morally the same thing as a Tensor. We expended a lot of effort to unify them.

So, we unified them in PyTorch proper, why not here? I don't think there's necessarily anything to do here, but I would like to understand better why wrapping was the right thing to do here.

Reply:
> Although, the concern there was that we shouldn't do that since RPC messages could be used for a variety of different purposes and in many cases RPC messages won't have any autograd information associated with them.

Err... so what other uses are we going to use the RPC messages for?

Reply:
@ezyang 

> Err... so what other uses are we going to use the RPC messages for?

Any RPC that does not require autograd, e.g., ACK messages, running some user functions under no_grad mode, initialization/coordination UDFs. I described my thoughts on this in another comment below. I agree that we shouldn't double wrap a message, but I think we don't need to add an extra field to Message. Currently what we have is:

* On sender:

context-aware messages (e.g., `ScriptCall`) -> context-agnostic message (`Message`) -> `RpcAgent::send(Message)`

* On receiver:

`RpcAgent::receiveLoop` -> context-agnostic message (`Message`) -> context-aware messages (e.g., `ScriptCall`) -> `RequestCallback`.

My thought is that `Message` and `RpcAgent` should be pure communication layer thing, and should not care about the content in any messages. It is because we will have different implementations of `RpcAgent` using different underlying substrate. And it does not make sense to force all of them to understand distributed autograd. The distributed autograd logic should only be implemented once and work for all `RpcAgent` implementations. However, this does not mean we have to double wrap a message either. Those context-related logic can be implemented in context-aware messages, e.g., in `ScriptCall::toMessage`, `ScriptCall::fromMessage`, and its `RequestCallback`.

Reply:
@mrshenli and I discussed this over lunch yesterday. @mrshenli, do you think you can reproduce the discussion here?

Reply:
Sure, IIUC, two things were mentioned in the discussion:

1. If we consider `ScriptCall` as a data layer message and `Message` as a comm layer message, then it looks better to add autograd meta info as a field to data layer messages instead of creating a new data layer message that wraps another data layer message.
  
   > It seems really tedious to have multiple classes like SCRIPT_CALL_WITH_GRAD, SCRIPT_RET_WITH_GRAD etc.
  
   It does not has to be using dedicated message types for them. Another option could be adding that autograd meta info as a field to the base class of all data layer message types. It is similar to what you are doing in this PR, but the difference is that we won't have double wrapping or dedicated autograd message types. And you don't have to manually setup bytes in the payload either. It could be, e.g., `ScriptCall::toMessage` calls `RpcCommandBase::toMessage` where the later returns a vector of IValues. Then `ScriptCall` use jit pickle to serialize its own IValues and the ones it got from `RpcCommandBase` in one shot. 
  
2. @ezyang mentioned that we could avoid adding fields to existing RPC messages if we have dedicated per-tensor send/recv messages. We haven't talked through all details on how to do this. (Below are my thoughts. @ezyang correct me if I am wrong.) Giving it a second thought, if we use dedicated send/recv messages, it would turn each RPC into multiple messages and we have to reassemble them on the other end. It might hurt perf?

Reply:
It seems like we're going back to the original design I had in this PR (with slight modifications). The reason we moved to the design we have in this PR is that the RPC layer should be agnostic about autograd. What is the reason we are moving away from this? When I thought about it a few weeks back it made a lot of sense to keep the autograd information outside of the RPC layer. The RPC layer can be used in a variety of ways where we don't have anything to do with autograd and as a result we shouldn't have any autograd logic in the RPC layer.

>I agree with @ezyang here. As soon as a message has a member tensors(), that means that this message is responsible for sending tensors over the network. If any of these tensors has a requires_grad=True, then it becomes responsibility of the message to handle that. It seems that here, this responsibility translates into injecting this AutogradMetadata into the message.

> Here, the base class "Message" exposes the member tensors(), and I believe that this is a strong enough evidence that messages should handle the autogradn'ess of those tensors its' transmitting.

This isn't entirely true. Even if the tensors have `requires_grad=True` it doesn't mean we need to handle autograd, since we may not have a valid autograd context.